### PR TITLE
DR-3180 Enable secure monitoring on existing GCP datasets and their associated snapshots

### PR DESCRIPTION
**Pull request by snf2ye**
### Testing

Changes are deployed to my dev environment (https://jade-sh.datarepo-dev.broadinstitute.org/). You can call it with the following request or use the swagger page:

```
curl -X 'POST' \
  'https://jade-sh.datarepo-dev.broadinstitute.org/api/repository/v1/upgrade' \
  -H 'accept: application/json' \
  -H 'Authorization: Bearer xxx' \
  -H 'Content-Type: application/json' \
  -d '{
  "upgradeName": "ENABLE_SECURE_MONITORING",
  "upgradeType": "custom",
  "customName": "ENABLE_SECURE_MONITORING",
  "customArgs": [
    "<your dataset id>"
  ]
}'
```

### Tasks

- [x] Add new upgrade/migration flight
- [x] Add step to enable secure monitoring flag
- [x] Add test for secure monitoring flag
- [x] Add step to refolder dataset & child snapshot's GCP project on enabling secure monitoring
- [x] Add test for refoldering projects
- [x] Add step to create TPS policies 
- [x] Test flight manually and confirm each resource/step is correctly executed
- [x] Bug: If TPS policy already exists for snapshot, the flight fails.

### CODE REVIEW

File: src/main/java/bio/terra/service/dataset/DatasetService.java
Comment (okotsopoulos) [inline src/main/java/bio/terra/service/dataset/DatasetService.java]: Idea -- maybe we should only support turning secure monitoring on?  I know that would make the patch DAO method a bit more complex.

Comment (snf2ye) [reply src/main/java/bio/terra/service/dataset/DatasetService.java]: I think that'd be reasonable, especially for this quick turnaround.

File: src/main/java/bio/terra/service/dataset/DatasetService.java
Comment (okotsopoulos) [inline src/main/java/bio/terra/service/dataset/DatasetService.java]: Is there any harm to running this method if the dataset already had secure monitoring enabled?

File: src/main/java/bio/terra/service/dataset/DatasetService.java
Comment (okotsopoulos) [inline src/main/java/bio/terra/service/dataset/DatasetService.java]: Do we have a sense of how long this process takes?  The patch endpoint returns synchronously so could be nice to have a rough idea.

File: src/main/java/bio/terra/service/dataset/DatasetService.java
Comment (snf2ye) [reply src/main/java/bio/terra/service/dataset/DatasetService.java]: Very good question -- it's very possible but not certain that the refoldering request could fail.

Comment (okotsopoulos) [reply src/main/java/bio/terra/service/dataset/DatasetService.java]: Cool.  A failure isn't necessarily a bad thing here, as long as it doesn't leave things in a bad state.  We don't have access to resource locking as part of the patch endpoint which is how we might handle this in a flight (not arguing for a flight, I think it might be overkill given the context of this request).

File: src/main/java/bio/terra/service/dataset/DatasetDao.java
Comment (pshapiro4broad) [inline src/main/java/bio/terra/service/dataset/DatasetDao.java]:
```suggestion
        "UPDATE dataset SET secure_monitoring = :enabledSecureMonitoring WHERE id = :id";
```

File: src/main/java/bio/terra/service/upgrade/UpgradeService.java
Comment (pshapiro4broad) [inline src/main/java/bio/terra/service/upgrade/UpgradeService.java:44]: I'm not familiar with `Preconditions` but if you had a function like it but handled `Optional<T>` you could instead write something like:
```
UUID datasetId = Preconditions.optionalPresent(ValidationUtils.convertToUuid(request.getCustomArgs().get(0),
    "Custom argument's single value is not a valid UUID");
return Map.of(JobMapKeys.DATASET_ID.getKeyName(), datasetId);
```
The goal being to avoid splitting up an `isPresent()` check from its corresponding `get()`.

File: src/main/java/bio/terra/service/dataset/DatasetService.java
Comment (snf2ye) [reply src/main/java/bio/terra/service/dataset/DatasetService.java]: The new approach only supports enabling secure monitoring.

File: src/main/java/bio/terra/service/dataset/DatasetService.java
Comment (snf2ye) [reply src/main/java/bio/terra/service/dataset/DatasetService.java]: I added a bit of logic to skip the project if it is already in the secure monitoring folder. I've also added to a test to confirm that this works okay!

File: src/main/java/bio/terra/service/dataset/DatasetService.java
Comment (snf2ye) [reply src/main/java/bio/terra/service/dataset/DatasetService.java]: Very good point, I'm not sure how long it would take. Hopefully moving this to a flight should handle this case, too.

File: src/main/java/bio/terra/service/dataset/DatasetService.java
Comment (okotsopoulos) [reply src/main/java/bio/terra/service/dataset/DatasetService.java]: Awesome!

File: src/main/java/bio/terra/service/dataset/DatasetService.java
Comment (okotsopoulos) [reply src/main/java/bio/terra/service/dataset/DatasetService.java]: Agreed!

File: src/main/java/bio/terra/service/dataset/flight/upgrade/enableSecureMonitoring/EnableSecureMonitoringCreateTpsPolicyStep.java
Comment (okotsopoulos) [inline src/main/java/bio/terra/service/dataset/flight/upgrade/enableSecureMonitoring/EnableSecureMonitoringCreateTpsPolicyStep.java]: Not mandatory, but would it be preferable to query `snapshot_source` directly for this?

```
datarepo=# \d snapshot_source;
                Table "public.snapshot_source"
   Column    | Type | Collation | Nullable |      Default      
-------------+------+-----------+----------+-------------------
 id          | uuid |           | not null | gen_random_uuid()
 snapshot_id | uuid |           | not null | 
 dataset_id  | uuid |           | not null | 
 asset_id    | uuid |           |          | 
Indexes:
    "snapshot_source_pkey" PRIMARY KEY, btree (id)
Foreign-key constraints:
    "fk_snapshot_source_asset_specification" FOREIGN KEY (asset_id) REFERENCES asset_specification(id) ON DELETE CASCADE
    "fk_snapshot_source_dataset" FOREIGN KEY (dataset_id) REFERENCES dataset(id) ON DELETE CASCADE
    "fk_snapshot_source_snapshot" FOREIGN KEY (snapshot_id) REFERENCES snapshot(id) ON DELETE CASCADE
Referenced by:
    TABLE "snapshot_map_table" CONSTRAINT "fk_snapshot_map_snapshot_source" FOREIGN KEY (source_id) REFERENCES snapshot_source(id) ON DELETE CASCADE
```

File: src/main/java/bio/terra/service/dataset/flight/upgrade/enableSecureMonitoring/EnableSecureMonitoringCreateTpsPolicyStep.java
Comment (okotsopoulos) [inline src/main/java/bio/terra/service/dataset/flight/upgrade/enableSecureMonitoring/EnableSecureMonitoringCreateTpsPolicyStep.java]: That's a good question. I am inclined towards not deleting the PAOs if there's a chance they already existed before the flight.

And if this change were made, I think that retriggering a flight that failed midway through creating snapshot PAOs would be able to progress (i.e. the flight should not fail if we try to create a snapshot PAO that already exists, which it looks like you've handled above).

File: src/main/java/bio/terra/service/dataset/flight/upgrade/enableSecureMonitoring/EnableSecureMonitoringFlight.java
Comment (okotsopoulos) [inline src/main/java/bio/terra/service/dataset/flight/upgrade/enableSecureMonitoring/EnableSecureMonitoringFlight.java]: Thanks for adding this conditional handling!

File: src/main/java/bio/terra/service/dataset/flight/upgrade/enableSecureMonitoring/EnableSecureMonitoringFlight.java
Comment (okotsopoulos) [inline src/main/java/bio/terra/service/dataset/flight/upgrade/enableSecureMonitoring/EnableSecureMonitoringFlight.java:43]: Nice, I agree with obtaining an exclusive lock on the dataset.

File: src/main/java/bio/terra/service/dataset/flight/upgrade/enableSecureMonitoring/EnableSecureMonitoringFlipFlagStep.java
Comment (okotsopoulos) [inline src/main/java/bio/terra/service/dataset/flight/upgrade/enableSecureMonitoring/EnableSecureMonitoringFlipFlagStep.java]: Calling out an edge case:

Presently I think it's possible that this flight could be called on a dataset that already has secure monitoring enabled.  If this step were undone, then we might inadvertently set the flag to false.

Should we conditionally run the flight's steps only when secure monitoring is not already enabled?
Or, should we record the prior secure monitoring flag value to the working map and pass it to the DAO update method instead?

File: src/main/java/bio/terra/service/dataset/flight/upgrade/enableSecureMonitoring/EnableSecureMonitoringRefolderGCPProjectsStep.java
Comment (okotsopoulos) [inline src/main/java/bio/terra/service/dataset/flight/upgrade/enableSecureMonitoring/EnableSecureMonitoringRefolderGCPProjectsStep.java]: Naming nit -- Google Java style guide advises camel case for acronyms, so this would be `EnableSecureMonitoringRefolderGcpProjectsStep`.

https://google.github.io/styleguide/javaguide.html#s5.3-camel-case

File: src/main/java/bio/terra/service/dataset/flight/upgrade/enableSecureMonitoring/EnableSecureMonitoringRefolderGCPProjectsStep.java
Comment (okotsopoulos) [inline src/main/java/bio/terra/service/dataset/flight/upgrade/enableSecureMonitoring/EnableSecureMonitoringRefolderGCPProjectsStep.java]: I appreciate this comment!

File: src/main/java/bio/terra/service/dataset/flight/upgrade/enableSecureMonitoring/EnableSecureMonitoringFlight.java
Comment (okotsopoulos) [inline src/main/java/bio/terra/service/dataset/flight/upgrade/enableSecureMonitoring/EnableSecureMonitoringFlight.java]: I had to dig into the code a bit before I realized that this step was handling TPS policy creation for all of the dataset's snapshots.  Could be worth a rename to the step to mention the snapshot part (I know it's a long name already), and/or adding comments inline to this flight construction.

File: src/main/java/bio/terra/service/snapshot/SnapshotService.java
Comment (okotsopoulos) [inline src/main/java/bio/terra/service/snapshot/SnapshotService.java]: This could be consolidated as `.map(SnapshotSummaryModel::getDataProject).toList()`.

**Review by okotsopoulos — COMMENTED** [review]
Thank you so much for taking on this work.

I have a few high-level questions:

1. I can't quite remember who can call these upgrade flights -- is it only TDR admins?  Do you have a sense of what permission checking is performed, as well as what permissions on the underlying dataset and snapshots are required to run this?

2. I think that all the affected AnVIL datasets and snapshots are self-hosted.  Do you know if any special considerations are needed for the self-hosted case?  Can we manually test that as well?

File: src/main/java/bio/terra/service/dataset/flight/upgrade/enableSecureMonitoring/EnableSecureMonitoringCreateTpsPolicyStep.java
Comment (snf2ye) [reply src/main/java/bio/terra/service/dataset/flight/upgrade/enableSecureMonitoring/EnableSecureMonitoringCreateTpsPolicyStep.java]: We're on the same wavelength - I initially wrote a direct snapshot_source query and then decided to just use an existing query 😄  

I think the performance gains here only really matter if we make this permanent code -- Is that correct? 
Given the temporary nature of this code along with the reliability of using a well-used method, I chose the existing method.

**Comment by snf2ye** [comment]
> 1. I can't quite remember who can call these upgrade flights -- is it only TDR admins?  Do you have a sense of what permission checking is performed, as well as what permissions on the underlying dataset and snapshots are required to run this?

@okotsopoulos  - Excellent, excellent question. Thanks for asking because I hadn't thought about permissions yet.
But, yes, you have to be an admin to run an upgrade flight.
image

File: src/main/java/bio/terra/service/dataset/flight/upgrade/enableSecureMonitoring/EnableSecureMonitoringFlight.java
Comment (snf2ye) [reply src/main/java/bio/terra/service/dataset/flight/upgrade/enableSecureMonitoring/EnableSecureMonitoringFlight.java]: Eh, they're already annoyingly long names anyway. I've updated it - let me know if you another name suggestion!

File: src/main/java/bio/terra/service/dataset/flight/upgrade/enableSecureMonitoring/EnableSecureMonitoringFlipFlagStep.java
Comment (snf2ye) [reply src/main/java/bio/terra/service/dataset/flight/upgrade/enableSecureMonitoring/EnableSecureMonitoringFlipFlagStep.java]: Oh, that's a great point. I'll think about these approaches.

File: src/main/java/bio/terra/service/dataset/flight/upgrade/enableSecureMonitoring/EnableSecureMonitoringCreateSourceDatasetAndSnapshotsTpsPolicyStep.java
Comment (nmalfroy) [inline src/main/java/bio/terra/service/dataset/flight/upgrade/enableSecureMonitoring/EnableSecureMonitoringCreateSourceDatasetAndSnapshotsTpsPolicyStep.java:51]: We should have a log message here to let us know what snapshots are getting TPS policies

File: src/main/java/bio/terra/service/dataset/flight/upgrade/enableSecureMonitoring/EnableSecureMonitoringCreateSourceDatasetAndSnapshotsTpsPolicyStep.java
Comment (nmalfroy) [inline src/main/java/bio/terra/service/dataset/flight/upgrade/enableSecureMonitoring/EnableSecureMonitoringCreateSourceDatasetAndSnapshotsTpsPolicyStep.java]: You can just to a `toList()` here I think?

File: src/main/java/bio/terra/service/dataset/flight/upgrade/enableSecureMonitoring/EnableSecureMonitoringFlight.java
Comment (nmalfroy) [inline src/main/java/bio/terra/service/dataset/flight/upgrade/enableSecureMonitoring/EnableSecureMonitoringFlight.java]: This isn't strictly true.  It's supported on Azure but we don't want to add the patch  to set it to true

File: src/main/java/bio/terra/service/dataset/flight/upgrade/enableSecureMonitoring/EnableSecureMonitoringFlipFlagStep.java
Comment (nmalfroy) [inline src/main/java/bio/terra/service/dataset/flight/upgrade/enableSecureMonitoring/EnableSecureMonitoringFlipFlagStep.java]: FlipFlag.  That's fun to say :)

File: src/main/java/bio/terra/service/dataset/flight/upgrade/enableSecureMonitoring/EnableSecureMonitoringFlipFlagStep.java
Comment (nmalfroy) [reply src/main/java/bio/terra/service/dataset/flight/upgrade/enableSecureMonitoring/EnableSecureMonitoringFlipFlagStep.java]: I like the pattern of recording the starting state and then the undo is just to set it to the original state

File: src/main/java/bio/terra/service/upgrade/UpgradeService.java
Comment (nmalfroy) [reply src/main/java/bio/terra/service/upgrade/UpgradeService.java:44]: Oh need, I didn't realize preconditions had an optional handler built-in.

**Review by nmalfroy — COMMENTED** [review]
This looks great!  Thanks for jumping on it!  I think the main thing is to check for accidental rollback

File: src/main/java/bio/terra/service/upgrade/UpgradeService.java
Comment (pshapiro4broad) [reply src/main/java/bio/terra/service/upgrade/UpgradeService.java:44]: It's not built-in, we'd have to write it. But it would be easy to do and improve the code IMO

Comment (pshapiro4broad) [reply src/main/java/bio/terra/service/upgrade/UpgradeService.java:44]: Something like
```
  static <T> T optionalPresent(Optional<T> optional, String message) {
    return optional.orElseThrow(() -> new IllegalArgumentException(message));
  }
```

File: src/main/java/bio/terra/service/dataset/flight/upgrade/enableSecureMonitoring/EnableSecureMonitoringCreateSourceDatasetAndSnapshotsTpsPolicyStep.java
Comment (snf2ye) [reply src/main/java/bio/terra/service/dataset/flight/upgrade/enableSecureMonitoring/EnableSecureMonitoringCreateSourceDatasetAndSnapshotsTpsPolicyStep.java]: Updated! (Also, I think sonarcloud's "code smells" caught this as well, which is neat!)

File: src/main/java/bio/terra/service/dataset/flight/upgrade/enableSecureMonitoring/EnableSecureMonitoringFlight.java
Comment (snf2ye) [reply src/main/java/bio/terra/service/dataset/flight/upgrade/enableSecureMonitoring/EnableSecureMonitoringFlight.java]: True - I've added a little more nuance to this message.

File: src/main/java/bio/terra/service/dataset/flight/upgrade/enableSecureMonitoring/EnableSecureMonitoringFlight.java
Comment (okotsopoulos) [inline src/main/java/bio/terra/service/dataset/flight/upgrade/enableSecureMonitoring/EnableSecureMonitoringFlight.java:53]: Aha!  I just realized, adding a Journal entry recording the switch to secure logging could be a great addition here, even just for the source dataset (probably more fuss than it's worth to journal on each modified snapshot).  That way there will be a record in a dataset's activity log of exactly when secure logging was turned on.

File: src/main/java/bio/terra/service/upgrade/UpgradeService.java
Comment (pshapiro4broad) [reply src/main/java/bio/terra/service/upgrade/UpgradeService.java:44]: Not to drag this out too much but this is IMO a case where `Optional` is just getting in the way. If `convertToUuid()` returned `UUID` or `null` then you could write:
```
          UUID datasetId =
              Preconditions.checkNotNull(
                  ValidationUtils.convertToUuid(request.getCustomArgs().get(0)),
                  "Custom argument's single value is not a valid UUID");
```

Comment (snf2ye) [reply src/main/java/bio/terra/service/upgrade/UpgradeService.java:44]: This chunk of code is following the existing pattern so I'd like to keep it for this time-sensitive PR. But, I'll create a follow-on to update the logic! 
https://broadworkbench.atlassian.net/browse/DR-3184

File: src/main/java/bio/terra/service/dataset/flight/upgrade/enableSecureMonitoring/EnableSecureMonitoringFlight.java
Comment (snf2ye) [reply src/main/java/bio/terra/service/dataset/flight/upgrade/enableSecureMonitoring/EnableSecureMonitoringFlight.java:53]: There already is one on flipping the flag! [code change here
](https://github.com/DataBiosphere/jade-data-repo/pull/1498/files#diff-a534083fafcf74fa6c64aabdd30c69d5cfc96e296f774302299bf2aa0a4884e8R760)

Do you think that's sufficient?

File: src/main/java/bio/terra/service/dataset/flight/upgrade/enableSecureMonitoring/EnableSecureMonitoringCreateTpsPolicyStep.java
Comment (okotsopoulos) [reply src/main/java/bio/terra/service/dataset/flight/upgrade/enableSecureMonitoring/EnableSecureMonitoringCreateTpsPolicyStep.java]: Great reasoning.

File: src/main/java/bio/terra/service/dataset/flight/upgrade/enableSecureMonitoring/EnableSecureMonitoringFlight.java
Comment (okotsopoulos) [reply src/main/java/bio/terra/service/dataset/flight/upgrade/enableSecureMonitoring/EnableSecureMonitoringFlight.java:53]: Oh neat, I missed that!

I think especially given the time-sensitivity this is great.  In some other flights we've set up the journaling so that if the step is rolled back, the journal entry written is removed.  But at least here if the step is rolled back we will see two journal entries: one indicating that monitoring was enabled and one indicating that monitoring is disabled, so that's just fine.

Comment (snf2ye) [reply src/main/java/bio/terra/service/dataset/flight/upgrade/enableSecureMonitoring/EnableSecureMonitoringFlight.java:53]: Actually, I think you have the right idea. I'm moving it to its own step at the end of the flight.

**Review by okotsopoulos — APPROVED** [review]
Looks awesome!

File: src/main/java/bio/terra/service/dataset/flight/upgrade/enableSecureMonitoring/EnableSecureMonitoringCreateSourceDatasetAndSnapshotsTpsPolicyStep.java
Comment (pshapiro4broad) [inline src/main/java/bio/terra/service/dataset/flight/upgrade/enableSecureMonitoring/EnableSecureMonitoringCreateSourceDatasetAndSnapshotsTpsPolicyStep.java:51]: Is this operation idempotent, if the flight is restarted? It looks like the underlying API in TPS might return a `409` if a policy for that ID already exists.

Comment (snf2ye) [reply src/main/java/bio/terra/service/dataset/flight/upgrade/enableSecureMonitoring/EnableSecureMonitoringCreateSourceDatasetAndSnapshotsTpsPolicyStep.java:51]: Yes - We catch the conflict exception so the flight should complete regardless of whether the policy was added. I've tested the case where the policy already exists.

File: src/main/java/bio/terra/service/dataset/flight/upgrade/enableSecureMonitoring/EnableSecureMonitoringEnableFlagStep.java
Comment (samanehsan) [inline src/main/java/bio/terra/service/dataset/flight/upgrade/enableSecureMonitoring/EnableSecureMonitoringEnableFlagStep.java:27]: Would it make sense to only do this patch if `DatasetWorkingMapKeys.SECURE_MONITORING_ENABLED` is false?

File: src/main/java/bio/terra/service/dataset/flight/upgrade/enableSecureMonitoring/EnableSecureMonitoringRefolderGcpProjectsStep.java
Comment (samanehsan) [inline src/main/java/bio/terra/service/dataset/flight/upgrade/enableSecureMonitoring/EnableSecureMonitoringRefolderGcpProjectsStep.java:57]: Was `snapshotService.enumerateGcpProjectsForSourceDataset` meant to replace this?

Comment (snf2ye) [reply src/main/java/bio/terra/service/dataset/flight/upgrade/enableSecureMonitoring/EnableSecureMonitoringRefolderGcpProjectsStep.java:57]: Ah, shoot. Thanks for catching that. I thought I had removed that method from snapshot service. (and I'll do that again!)
 Since this method is only used in this flight and is potentially temporary code, I thought I would just add the helper method to the step code.

File: src/main/java/bio/terra/service/dataset/flight/upgrade/enableSecureMonitoring/EnableSecureMonitoringEnableFlagStep.java
Comment (snf2ye) [reply src/main/java/bio/terra/service/dataset/flight/upgrade/enableSecureMonitoring/EnableSecureMonitoringEnableFlagStep.java:27]: That would definitely also be a valid approach. 
I think the use case y'all discussed at standup could make it worth it to keep the upgrade even if it's already true -- It will still go through and create the TPS policies.

File: src/main/java/bio/terra/service/dataset/flight/upgrade/enableSecureMonitoring/EnableSecureMonitoringRefolderGcpProjectsStep.java
Comment (samanehsan) [reply src/main/java/bio/terra/service/dataset/flight/upgrade/enableSecureMonitoring/EnableSecureMonitoringRefolderGcpProjectsStep.java:57]: That makes sense! Thanks for clarifying!

File: src/main/java/bio/terra/service/dataset/flight/upgrade/enableSecureMonitoring/EnableSecureMonitoringEnableFlagStep.java
Comment (samanehsan) [reply src/main/java/bio/terra/service/dataset/flight/upgrade/enableSecureMonitoring/EnableSecureMonitoringEnableFlagStep.java:27]: Oh I was referring to just this step. So if we know that a dataset already has secure monitoring enabled, we could skip this call to `datasetDao.setSecureMonitoring`. Not worth blocking the PR though!

Comment (snf2ye) [reply src/main/java/bio/terra/service/dataset/flight/upgrade/enableSecureMonitoring/EnableSecureMonitoringEnableFlagStep.java:27]: That is an excellent point. 
But, yeah, we're getting close to PR blocking territory. I've already done manual testing and I'd like for automated tests to run before merging today. And, while not optimal, it still should be okay without this change.

File: src/main/java/bio/terra/service/dataset/flight/upgrade/enableSecureMonitoring/EnableSecureMonitoringCreateSourceDatasetAndSnapshotsTpsPolicyStep.java
Comment (pshapiro4broad) [reply src/main/java/bio/terra/service/dataset/flight/upgrade/enableSecureMonitoring/EnableSecureMonitoringCreateSourceDatasetAndSnapshotsTpsPolicyStep.java:51]: Doh, of course I see the catch right here... I was thinking it was outside the loop but it's inside. Never mind.

**Review by samanehsan — APPROVED** [review]
Thanks @snf2ye!

File: src/main/java/bio/terra/service/dataset/flight/upgrade/enableSecureMonitoring/EnableSecureMonitoringFlipFlagStep.java
Comment (snf2ye) [reply src/main/java/bio/terra/service/dataset/flight/upgrade/enableSecureMonitoring/EnableSecureMonitoringFlipFlagStep.java]: I renamed it 😢

**Pull request merged**
